protocols/relay/src/handler.rs

```diff
@@ -0,0 +1,550 @@
+    incoming_dst_req_pending_approval: HashMap<RequestId, protocol::IncomingDstReq>,
+    /// Futures that send back an accept response to a relay.
+    accept_dst_futures: FuturesUnordered<
+        BoxFuture<
+            'static,
+            Result<
+                (PeerId, protocol::Connection, oneshot::Receiver<()>),
+                protocol::IncomingDstReqError,
+            >,
+        >,
+    >,
+    /// Futures that copy from a source to a destination.
+    copy_futures: FuturesUnordered<BoxFuture<'static, Result<(), protocol::IncomingRelayReqError>>>,
+    /// Requests asking the remote to become a relay.
+    outgoing_relay_reqs: Vec<OutgoingRelayReq>,
+    /// Requests asking the remote to become a destination.
+    outgoing_dst_reqs: SmallVec<[(PeerId, RequestId, Multiaddr, protocol::IncomingRelayReq); 4]>,
+    /// Queue of events to return when polled.
+    queued_events: Vec<RelayHandlerEvent>,
+    /// Tracks substreams lend out to other [`RelayHandler`]s or as
+    /// [`Connection`](protocol::Connection) to the
+    /// [`RelayTransportWrapper`](crate::RelayTransportWrapper).
+    ///
+    /// For each substream to the peer of this handler, there is a future in here that resolves once
+    /// the given substream is dropped.
+    ///
+    /// Once all substreams are dropped and this handler has no other work, [`KeepAlive::Until`] can
+    /// be set eventually allowing the connection to be closed.
+    alive_lend_out_substreams: FuturesUnordered<oneshot::Receiver<()>>,
+    /// The current connection keep-alive.
+    keep_alive: KeepAlive,
+}
+
+struct OutgoingRelayReq {
+    src_peer_id: PeerId,
+    dst_peer_id: PeerId,
+    request_id: RequestId,
+    /// Addresses of the destination.
+    dst_addr: Multiaddr,
+}
+
+/// Event produced by the relay handler.
+pub enum RelayHandlerEvent {
+    /// The remote wants us to relay communications to a third party. You must either send back a
+    /// `DenyIncomingRelayReq`, or send a `OutgoingDstReq` to a different handler containing
+    /// this object.
+    IncomingRelayReq {
+        request_id: RequestId,
+        src_addr: Multiaddr,
+        req: protocol::IncomingRelayReq,
+    },
+
+    /// The remote is a relay and is relaying a connection to us. In other words, we are used as
+    /// destination. The behaviour can accept or deny the request via
+    /// [`AcceptDstReq`](RelayHandlerIn::AcceptDstReq) or
+    /// [`DenyDstReq`](RelayHandlerIn::DenyDstReq).
+    IncomingDstReq(PeerId, RequestId),
+
+    /// A `RelayReq` that has previously been sent has been accepted by the remote. Contains
+    /// a substream that communicates with the requested destination.
+    ///
+    /// > **Note**: There is no proof that we are actually communicating with the destination. An
+    /// >           encryption handshake has to be performed on top of this substream in order to
+    /// >           avoid MITM attacks.
```

Comment:
First of, this note is from previous implementation attempts (https://github.com/libp2p/rust-libp2p/pull/1134/files#diff-3e891c993e81968e6593c3e9dd2806b688f4b0b30d30c246ac63f016a86e2df2R87), which this work is based on, thus I might have a wrong understanding myself.

> The relay spec says "Relayed connections are end-to-end encrypted just like regular connections." but I don't see how that is achieved with this protocol, as the relay node can read all traffic.

The relay protocol itself does not secure (authenticate / encrypt / ...) relayed connections. But given that relayed connections are treated just like any other connection by the source and destination node, they too go through the configured encryption and multiplexing upgrade.

> Well, by nature of this protocol, one usually trusts the relay node

With the above in mind and under the circumstance that the source is aware of the destinations `PeerId` (i.e. public key), not just its `Multiaddr`, I would claim that the source does not need to trust the relay.

@romanb does the above make sense? If so, how could we improve the comment to avoid future confusions? I might be missing something here.

Reply:
> But given that relayed connections are treated just like any other connection by the source and destination node, they too go through the configured encryption and multiplexing upgrade.

It may just have been a misunderstanding on my part. So, to clarify, a "relayed connection" between A and B through R consists of two substreams, one between A and R and one between R and B. These substreams exist on fully upgraded, authenticated, multiplexed (TCP) connections between A and R, as well as R and B, respectively. These two substreams form the relayed "connection", and they are the only substreams on the respective connections. This relayed "connection" comprising these two substreams then goes through all transport upgrades, authentication and multiplexing. So we have "nested" encryption and multiplexing. Is that right? Then indeed the relayed "connection" would be end-to-end encrypted (and authenticated) and I simply did not consider the second, nested transport upgrade happening through the negotiated substreams.

Reply:
> Is that right?

Yes, all of the above is correct, except:

> These substreams exist on fully upgraded, **authenticated**, multiplexed (TCP) connections between A and R, as well as R and B, respectively.

I guess one could as well use `libp2p-plaintext` for (non-) _authentication_ for the outer (TCP) connection in case leaking meta-data is ok. Though I have not seen any mention of this idea before and in addition `libp2p-plaintext` is explicitly designed for testing only.

> These two substreams form the relayed "connection", and **they are the only substreams on the respective connections**.

I am not aware of the constraint that the two TCP connections between A-R and R-B can at most carry a single relay substream. This implementation would potentially re-use the TCP connections in case additional relay requests could make use of either or both TCP connections. One could enforce the single TCP connection per relay connection constrained e.g. to minimize the impact of nested multiplexing. Enforcing this constrained would be difficult for destination nodes behind NATs, as a relay node would not be able to establish a new TCP connection for an incoming relay request to the destination. In addition the `Relay` behaviour might be used in conjunction with other `NetworkBehaviour`s, which themselves could potentially request substreams for the given connection.

Let me know whether the above makes sense @romanb.